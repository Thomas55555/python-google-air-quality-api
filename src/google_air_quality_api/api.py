"""API for Google Air Quality bound to Home Assistant OAuth."""

import logging
from .const import API_BASE_URL
from .auth import AbstractAuth
from .model import AirQualityData, UserInfoResult


_LOGGER = logging.getLogger(__name__)


USERINFO_API = "https://www.googleapis.com/oauth2/v1/userinfo"
CURRENT_CONDITIONS = f"{API_BASE_URL}airQuality/currentConditions"


class GoogleAirQualityApi:
    """The Google Photos library api client."""

    def __init__(self, auth: AbstractAuth) -> None:
        """Initialize GoogleAirQualityApi."""
        self._auth = auth

    async def async_air_quality(self, lat: float, long: float) -> AirQualityData:
        """Get all air quality data."""
        payload = {
            "location": {"latitude": lat, "longitude": long},
            "extraComputations": [
                "LOCAL_AQI",
                "POLLUTANT_CONCENTRATION",
            ],
            "customLocalAqis": [
                {"regionCode": "DE", "aqi": "USA_EPA_NOWCAST"},
            ],
            "universalAqi": True,
        }
        return await self._auth.post_json(
            CURRENT_CONDITIONS, json=payload, data_cls=AirQualityData
        )

    async def get_user_info(self) -> UserInfoResult:
        """Get the user profile info.

        This call requires the userinfo.email scope.
        """
        return await self._auth.get_json(USERINFO_API, data_cls=UserInfoResult)
