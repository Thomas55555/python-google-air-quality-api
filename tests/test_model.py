"""Snapshot tests for Google Air Quality API client."""

from dataclasses import fields

from syrupy.assertion import SnapshotAssertion
import pytest
from google_air_quality_api.model import AirQualityData, AQICategoryMapping
from typing import Any


def test_air_quality_snapshot(
    snapshot: SnapshotAssertion, air_quality_data: dict
) -> None:
    """Testing a snapshot of air quality data."""
    data = AirQualityData.from_dict(air_quality_data)

    # 1) Snapshot jedes einzelnen Feldes
    for field in fields(data):
        field_name = field.name
        field_value = getattr(data, field_name)
        assert field_value == snapshot(name=f"{field_name}")

    # 2) Zusätzlich: snapshotten der category_options aller Index-Einträge
    all_options = [idx.category_options for idx in data.indexes]
    assert all_options == snapshot(name="indexes_category_options")

    seen: dict[str, str] = {}
    for cat in AQICategoryMapping.get_all():
        if cat.normalized not in seen:
            seen[cat.normalized] = cat.original
    assert dict(sorted(seen.items())) == snapshot(name="all_aqi_categories")
