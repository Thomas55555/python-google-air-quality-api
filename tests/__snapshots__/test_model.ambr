# serializer version: 1
# name: test_air_quality_snapshot[_indexes]
  list([
    Index(code='uaqi', display_name='Universal AQI', color=Color(red=0.48235294, green=0.79607844, blue=0.2), category='good_air_quality', dominant_pollutant='o3', aqi=71, aqi_display='71'),
    Index(code='deu_uba', display_name='LQI (DE)', color=Color(red=0.3137255, green=0.8039216, blue=0.6666667), category='good_air_quality', dominant_pollutant='no2', aqi=None, aqi_display=None),
  ])
# ---
# name: test_air_quality_snapshot[_pollutants]
  list([
    Pollutant(code='co', display_name='CO', full_name='Carbon monoxide', concentration=Concentration(value=0.19197, units='ppm')),
    Pollutant(code='no2', display_name='NO2', full_name='Nitrogen dioxide', concentration=Concentration(value=16.97, units='ppb')),
    Pollutant(code='o3', display_name='O3', full_name='Ozone', concentration=Concentration(value=35.17, units='ppb')),
    Pollutant(code='pm10', display_name='PM10', full_name='Inhalable particulate matter (<10µm)', concentration=Concentration(value=13.53, units='µg/m³')),
    Pollutant(code='pm25', display_name='PM2.5', full_name='Fine particulate matter (<2.5µm)', concentration=Concentration(value=4.61, units='µg/m³')),
    Pollutant(code='so2', display_name='SO2', full_name='Sulfur dioxide', concentration=Concentration(value=1.37, units='ppb')),
  ])
# ---
# name: test_air_quality_snapshot[all_aqi_categories]
  dict({
    '10_33': '10-33% of guideline',
    '1_blue': '1 - Blue',
    '1_green': '1 - Green',
    '1a_very_good_air_quality': '1A - Very good air quality',
    '1b_good_air_quality': '1B - Good air quality',
    '2_cyan': '2 - Cyan',
    '2_light_green': '2 - Light green',
    '2_orange': '4 - Orange',
    '2_red': '5 - Red',
    '2_yellow': '3 - Yellow',
    '2a_acceptable_air_quality': '2A - Acceptable air quality',
    '2b_acceptable_air_quality': '2B - Acceptable air quality',
    '33_66': '33-66% of guideline',
    '3_green': '3 - Green',
    '3a_aggravated_air_quality': '3A - Aggravated air quality',
    '3b_bad_air_quality': '3B - Bad air quality',
    '4_yellow_watch': '4 - Yellow/Watch',
    '5_orange_alert': '5 - Orange/Alert',
    '66_100': '66-100% of guideline',
    '6_red_alert': '6 - Red/Alert+',
    'above_average_air_pollution': 'Above average air pollution',
    'acceptable': 'Acceptable',
    'acceptable_air_quality': 'Acceptable air quality',
    'acutely_unhealthy': 'Acutely unhealthy air quality',
    'alarm_level': 'Alarm level',
    'alert': 'Alert threshold',
    'alert_level': 'Alert level',
    'average_air_pollution': 'Average air pollution',
    'average_air_quality': 'Average air quality',
    'bad_air_quality': 'Bad air quality',
    'below_10': 'Less than 10% of guideline',
    'below_average_air_pollution': 'Below average air pollution',
    'caution': 'Caution',
    'clean': 'Clean',
    'considerable_air_pollution': 'Considerable air pollution',
    'degraded_air_quality': 'Degraded air quality',
    'desirable_air_quality': 'Desirable air quality',
    'emergency': 'Emergency',
    'emergency_level': 'Emergency level',
    'evident_air_pollution': 'Evident air pollution',
    'excellent': 'Excellent',
    'excellent_air_quality': 'Excellent air quality',
    'extremely_bad_air_quality': 'Extremely bad air quality',
    'extremely_poor_air_quality': 'Extremely poor air quality',
    'extremely_unfavorable_air_quality': 'Extremely unfavorable air quality',
    'extremely_unhealthy_air_quality': 'Extremely Unhealthy air quality',
    'fair_air_quality': 'Fair air quality',
    'fairly_good_air_quality': 'Fairly good air quality',
    'good': 'Good',
    'good_air_quality': 'Good air quality',
    'greater_100': 'Greater than 100% of guideline',
    'hazardous_air_quality': 'Hazardous air quality',
    'heavily_polluted': 'Heavily Polluted',
    'heavy_air_pollution': 'Heavy air pollution',
    'high_air_pollution': 'High air pollution',
    'high_air_quality': 'High air pollution',
    'high_health_risk': 'High health risk',
    'horrible_air_quality': 'Horrible air quality',
    'light_air_pollution': 'Light air pollution',
    'low': 'Low pollution',
    'low_air_pollution': 'Low air pollution',
    'low_air_quality': 'Low air quality',
    'low_health_risk': 'Low health risk',
    'medium_air_pollution': 'Medium air pollution',
    'medium_air_quality': 'Medium air quality',
    'moderate_air_pollution': 'Moderate air pollution',
    'moderate_air_quality': 'Moderate air quality',
    'moderate_health_risk': 'Moderate health risk',
    'moderately_polluted': 'Moderately polluted',
    'n1_good_air_quality': 'N1 - Good air quality',
    'n2_moderate_air_quality': 'N2 - Moderate air quality',
    'n3_bad_air_quality': 'N3 - Bad air quality',
    'n4_very_bad_air_quality': 'N4 - Very bad air quality',
    'n5_poor_air_quality': 'N5 - Poor air quality',
    'normal': 'Normal',
    'polluted': 'Polluted',
    'poor_air_quality': 'Poor air quality',
    'precautionary_level': 'Precautionary level',
    'reasonably_good_air_quality': 'Reasonably good air quality',
    'regular_air_quality': 'Regular air quality',
    'satisfactory_air_quality': 'Satisfactory air quality',
    'serious_air_pollution': 'Serious air pollution',
    'seriously_polluted': 'Seriously Polluted',
    'severe_air_pollution': 'Severe air pollution',
    'severe_air_quality': 'Severe air quality',
    'slightly_polluted': 'Slightly Polluted',
    'sufficient_air_quality': 'Sufficient air quality',
    'unfavorable_air_quality': 'Unfavorable air quality',
    'unfavorable_sensitive': 'Unfavorable air quality for sensitive groups',
    'unhealthy_air_quality': 'Unhealthy air quality',
    'unhealthy_sensitive': 'Unhealthy air quality for sensitive groups',
    'unsatisfactory_air_quality': 'Unsatisfactory air quality',
    'very_bad_air_quality': 'Very bad air quality',
    'very_good_air_quality': 'Very good air quality',
    'very_high_air_pollution': 'Very high air pollution',
    'very_high_air_quality': 'Very High air pollution',
    'very_high_health_risk': 'Very high health risk',
    'very_low_air_pollution': 'Very low air pollution',
    'very_polluted': 'Very Polluted',
    'very_poor_air_quality': 'Very poor air quality',
    'very_unfavorable_air_quality': 'Very unfavorable air quality',
    'very_unhealthy': 'Very unhealthy air quality',
    'very_unhealthy_air_quality': 'Very Unhealthy air quality',
    'warning_air_pollution': 'Warning level air pollution',
  })
# ---
# name: test_air_quality_snapshot[date_time]
  datetime.datetime(2025, 5, 25, 8, 0, tzinfo=datetime.timezone.utc)
# ---
# name: test_air_quality_snapshot[indexes_category_options]
  list([
    list([
      'excellent_air_quality',
      'good_air_quality',
      'moderate_air_quality',
      'low_air_quality',
      'poor_air_quality',
    ]),
    list([
      'very_good_air_quality',
      'good_air_quality',
      'moderate_air_quality',
      'poor_air_quality',
      'very_poor_air_quality',
    ]),
  ])
# ---
# name: test_air_quality_snapshot[indexes_pollutant_options]
  list([
    list([
      'co',
      'no2',
      'o3',
      'pm10',
      'pm25',
      'so2',
    ]),
    list([
      'no2',
      'o3',
      'pm10',
    ]),
  ])
# ---
# name: test_air_quality_snapshot[region_code]
  'de'
# ---
